kani-compiler/src/codegen_cprover_gotoc/codegen/foreign_function.rs

```diff
@@ -0,0 +1,155 @@
+    ///
+    /// CBMC built-in functions that are supported by Kani are always added to the symbol table, and
+    /// this function will return them.
+    ///
+    /// For other foreign items, we declare a shim and add to the list of foreign shims to be
+    /// handled later.
+    pub fn codegen_foreign_fn(&mut self, instance: Instance<'tcx>) -> &Symbol {
+        debug!(?instance, "codegen_foreign_function");
+        let fn_name = self.symbol_name(instance).intern();
+        if self.symbol_table.contains(fn_name) {
+            // Symbol has been added (either a built-in CBMC function or a Rust allocation function).
+            self.symbol_table.lookup(fn_name).unwrap()
+        } else if RUST_ALLOC_FNS.contains(&fn_name)
+            || (self.is_cffi_enabled() && kani_middle::fn_abi(self.tcx, instance).conv == Conv::C)
+        {
+            // Add a Rust alloc lib function as is declared by core.
+            // When C-FFI feature is enabled, we just trust the rust declaration.
+            // TODO: Add proper casting and clashing definitions check.
+            // https://github.com/model-checking/kani/issues/1350
+            // https://github.com/model-checking/kani/issues/2426
+            self.ensure(fn_name, |gcx, _| {
+                let typ = gcx.codegen_ffi_type(instance);
+                Symbol::function(
+                    fn_name,
+                    typ,
+                    None,
+                    gcx.readable_instance_name(instance),
+                    Location::none(),
+                )
+                .with_is_extern(true)
+            })
+        } else {
+            let shim_name = format!("{fn_name}_ffi_shim");
+            trace!(?shim_name, "codegen_foreign_function");
+            self.ensure(&shim_name, |gcx, _| {
+                // Generate a shim with an unsupported C-FFI error message.
+                let typ = gcx.codegen_ffi_type(instance);
+                Symbol::function(
+                    &shim_name,
+                    typ,
+                    Some(gcx.codegen_ffi_shim(shim_name.as_str().into(), instance)),
+                    gcx.readable_instance_name(instance),
+                    Location::none(),
+                )
+            })
+        }
+    }
+
+    /// Checks whether C-FFI has been enabled or not.
+    /// When enabled, we blindly encode the function type as is.
+    fn is_cffi_enabled(&self) -> bool {
+        self.queries.get_unstable_features().contains(&"c-ffi".to_string())
+    }
+
+    /// Generate code for a foreign function shim.
+    fn codegen_ffi_shim(&mut self, shim_name: InternedString, instance: Instance<'tcx>) -> Stmt {
+        debug!(?shim_name, ?instance, sym=?self.symbol_table.lookup(shim_name), "generate_foreign_shim");
+
+        let loc = self.codegen_span(&self.tcx.def_span(instance.def_id()));
+        let unsupported_check = self.codegen_ffi_unsupported(instance, loc);
+        Stmt::block(vec![unsupported_check], loc)
+    }
```

Comment:
Do we have a test case for this?

Reply:
yes. The new tests `expected/foreign-function/ffi_ptr.rs` and `cargo-kani/libc/` both exercise this code.